doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,264 @@
+  table consolidate out would be dependent on all the other tables
+
+### Option 2: Put all tables in a single persist "state"
+
+- Currently, persist has a 1:1 relationship between shards and the metadata
+  (batches, capabilities, etc) kept by the state machine. Instead, make the
+  metadata a map of `ShardId -> ShardMetadata`.
+- Pro: Atomic multi-table writes are trivial, as is advancing the upper of all
+  tables at once.
+- Con: Greatly adds complexity to the internals of persist, which is already
+  complex.
+- Con: Greatly complicates the API of persist.
+- Con: Membership changes (the mapping between a shard and which state it lives
+  in) are tricky.
+- Con: Significant risk of destabilizing persist, which has to work perfectly
+  for basically anything in Materialize to work.
+- Con: Reading or writing a single table would require loading the metadata of
+  all co-located shards, which scales arbitrarily large with the number of
+  tables. This likely implies an upper bound on the number of tables we can
+  support.
+- Con: Also likely includes a rewrite of how persist monitoring and debugging
+  tools work.
+
+### Option 3: Table remap shard
+
+- A single _table remap_ shard is introduced per environment, which maps _data
+  shard_ uppers to realtime.
+- Intuition: A data shard's upper is no longer the source of truth for when you
+  can read it. Instead, the remap shard is. This allows us to store uncommitted
+  data in the data shards.
+- The data shard is only written when new data arrives, _not_ as time passes.
+- Example:
+  - Before:
+
+        data0:
+            (Row1, T0-0, 1)
+            (Row2, T0-0, 1)
+            (Row3, T0-2, 1)
+            upper: 3
+        data1:
+            (Row4, T1-0, 1)
+            upper: 2
+        remap:
+            ((table0, T0-0), 2023-07-01, 1)
+            ((table0, T0-0), 2023-07-02, -1)
+            ((table0, T0-2), 2023-07-02, 1)
+            ((table1, T1-0), 2023-07-03, 1)
+
+  - First we write `(Row5, T1-1, 1)` to data1
+  - Then we write `((table1, T1-0), 2023-07-04, -1), ((table1, T1-1),
+    2023-07-04, 1)` to remap
+- This corresponds closely to the [reclocking doc], except that we're very
+  careful about how the `remap` collection is maintained to create the txn
+  behavior we want. We also (likely) don't write down the `reclocked` collection
+  itself, instead relying on reclocking being a definite computation of definite
+  inputs.
+- At-least-one somethings are responsible for repeatedly advancing the upper on
+  the remap shard. No coordination with table writes is necessary beyond that we
+  probably want to do something about contention.
+  - At the end of the day, there are two options for "do something about
+    contention": require the a txn's retry loop to finish in whatever time our
+    ts closing cadence is, or implement some form of holds with a timeout (and
+    probably heartbeats).
+  - Possible that this something wants to be co-located with the timestamp
```

Comment:
The current timestamp oracle (and also the one in my prototype) has the notion of acquiring a write timestamp and then "closing" it when the write completes. From this, you can derive when there are "open" transactions. Then, in my OCC prototype, I suspend periodic timestamp advancement when noticing this situation.

With this, anyone trying to promptly read things has a bad time. With the alternative (as you mentioned) where we periodically advance the remap timestamp anyone trying to do large-ish read-then-write transactions has a bad time. And, of course, any yolo write-transaction coming in while that slow read-then-write tries to complete is invalidating it (the read-then-write transaction). So I feel like either way we slice it, people trying to do large read-then-writes have a bad time. And this is probably a fine limitation to have. (cc @benesch and I talked about this before)